File: baekjoon/brute-force/python/BOJ6443.py
from sys import stdin
readline = stdin.readline

N = int(readline())

def solve():

    for _ in range(N):
        word = list(sorted(list(readline().strip())))
        print(''.join(word))
        while next_permutation(word):
            print(''.join(word))

def next_permutation(word):
    i = j = len(word) - 1

    while i > 0 and word[i - 1] >= word[i]:
        i -= 1

    if i == 0:
        return False
    
    while word[j] <= word[i - 1]:
        j -= 1

    word[i - 1], word[j] = word[j], word[i - 1]

    k = len(word) - 1

    while i < k:
        word[i], word[k] = word[k], word[i]
        i += 1
        k -= 1 

    return True

def swap(word, i, j):
    temp = word[i]
    word[i] = word[j]
    word[j] - temp

solve()

File: 0_1.baekjoon/brute-force/java/src/BOJ6443.java
import java.util.*;
import java.io.*;

public class BOJ6443 {
    public static void main(String[] args) throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        while (N-- > 0) {
            char[] word = br.readLine().strip().toCharArray();
            Arrays.sort(word);
            StringBuilder sb = new StringBuilder();
            sb.append(word).append("\n");

            while (nextPermutation(word)) {
                sb.append(word).append("\n");
            }

            System.out.print(sb);

        }
    }

    private static boolean nextPermutation(char[] word) {
        int i = word.length - 1, j = word.length - 1;

        while (i > 0 && word[i - 1] >= word[i]) {
            i--;
        }

        if (i == 0)
            return false;

        while (word[i - 1] >= word[j]) {
            j--;
        }

        swap(word, i - 1, j);
        
        Arrays.sort(word, i, word.length);

        return true;
    }

    private static void swap(char[] word, int i, int j) {
        char temp = word[i];
        word[i] = word[j];
        word[j] = temp;
    }
}
